### PR TITLE
Added initial set of tests for vsoClient

### PR DESCRIPTION
**Pull request by damccorm**
Divided into unit/non-unit tests, currently no non-unit tests added.

Addresses #217

### CODE REVIEW

**Comment by damccorm** [comment]
@stephenmichaelf I think I was pretty much able to cover this with L0 tests, did you have anything specific in mind that should be tested at L1 level? Or was that more for expanding this to other APIs in the future?

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -13,7 +13,13 @@ $ npm run build
 
 ## Test
 
-You should test with node 4.x, 6.x and 8.x LTS.  We recommend using nvm ([linux](https://github.com/creationix/nvm) / [windows](https://github.com/coreybutler/nvm-windows))
+You should test with node 4.x, 6.x and 8.x LTS.  We recommend using nvm ([linux](https://github.com/creationix/nvm) / [windows](https://github.com/coreybutler/nvm-windows)). To run units:
```
Comment (stephenmichaelf) [inline CONTRIBUTING.md]: Move "To run units:" to new line.

File: test/tests/vsoClientTests.ts [hunk 2]
```diff
@@ -0,0 +1,9 @@
+import assert = require('assert');
+import vsom = require('../../api/VsoClient');
+describe('VSOClient Tests', function () {
```
Comment (stephenmichaelf) [inline test/tests/vsoClientTests.ts]: Nit: add line between import and describe.

File: test/tests/vsoClientTests.ts [hunk 3]
```diff
@@ -0,0 +1,9 @@
+import assert = require('assert');
```
Comment (stephenmichaelf) [inline test/tests/vsoClientTests.ts]: Is this a placeholder? If we aren't using it for now we can remove until we need it.

File: test/units/vsoClientTests.ts [hunk 4]
```diff
@@ -0,0 +1,149 @@
+import assert = require('assert');
+import nock = require('nock');
+import vsom = require('../../api/VsoClient');
+import * as rm from 'typed-rest-client/RestClient';
+import { resolve } from 'path';
+describe('VSOClient Units', function () {
```
Comment (stephenmichaelf) [inline test/units/vsoClientTests.ts]: Nit: line between import and describe.

File: test/units/vsoClientTests.ts [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+    let rest: rm.RestClient;
+    let vsoClient: vsom.VsoClient
+    const baseUrl = 'https://dev.azure.com/';
+    before(() => {
+        const userAgent = "testAgent";
```
Comment (stephenmichaelf) [inline test/units/vsoClientTests.ts]: Add types.

File: test/units/vsoClientTests.ts [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+    after(() => {
+    });
+
+    afterEach(() => {
+        nock.cleanAll();
+        vsoClient = new vsom.VsoClient(baseUrl, rest);
+    });
+
+    it('constructs', () => {
+        this.timeout(1000);
```
Comment (stephenmichaelf) [inline test/units/vsoClientTests.ts]: Can we organize  these using Arrange, Act, Assert?(as commented sections e.g. // Arrange .... // Act ... // Assert

File: test/units/vsoClientTests.ts [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+        const userAgent = "testAgent";
+        const handlers = [];
+        const options = null;
+        const rest = new rm.RestClient(userAgent, null, handlers, options);
+        const vso: vsom.VsoClient = new vsom.VsoClient('https://microsoft.com', rest);
+        assert(vso, 'VsoClient should not be null');
+    });
+
+    it('gets versioning data where the max version and released version are identical', async () => {
+        nock('https://dev.azure.com/_apis/testArea', {
```
Comment (stephenmichaelf) [inline test/units/vsoClientTests.ts]: Spacing and Arrange Act Assert.

File: test/units/vsoClientTests.ts [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+            .reply(200, {
+                value: [{id: 'testLocation', maxVersion: '1', releasedVersion: '1', routeTemplate: 'testTemplate', area: 'testArea', resourceName: 'testName', resourceVersion: '1'}]
+               });
+        const res = await vsoClient.getVersioningData('1', 'testArea', 'testLocation', {'testKey': 'testValue'}, null);
+        assert(res.apiVersion === '1');
+        assert(res.requestUrl === 'https://dev.azure.com/testTemplate');
+    });
+
+    it('gets versioning data where ther requested version is greater than the max version', async () => {
+        nock('https://dev.azure.com/_apis/testArea2', {
```
Comment (stephenmichaelf) [inline test/units/vsoClientTests.ts]: Spacing and Arrange Act Assert.

Comment (stephenmichaelf) [reply test/units/vsoClientTests.ts]: Add for all tests.

**Comment by stephenmichaelf** [comment]
Are you now able to make PRs from the repo instead of your fork? Want to make sure permissions are correct.

File: make.js [hunk 9]
```diff
@@ -40,8 +40,22 @@ target.build = function() {
...
+    run('npm install ../_build');
+    popd();
+
+    console.log("-------Unit Tests-------");
+    run('tsc -p ./test/units');
+    run('mocha test/units');
+}
+
 target.test = function() {
+    target.units();
```
Comment (stephenmichaelf) [inline make.js]: In the previous version you had one more test here, did you mean to remove that too?

File: test/units/vsoClientTests.ts [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+import vsom = require('../../api/VsoClient');
+import * as rm from 'typed-rest-client/RestClient';
+import { resolve } from 'path';
+import { ApiResourceLocation } from '../../api/interfaces/common/VsoBaseInterfaces';
+
+describe('VSOClient Units', function () {
+    let rest: rm.RestClient;
+    let vsoClient: vsom.VsoClient
+    const baseUrl: string = 'https://dev.azure.com/';
+    before(() => {
```
Comment (stephenmichaelf) [inline test/units/vsoClientTests.ts]: Nit: add line between const and before.

File: make.js (hunk 9 above)
Comment (damccorm) [reply make.js]: Yeah, that was the test that had been a placeholder which I got rid of

**Comment by damccorm** [comment]
Yes, I am now able to make PRs from the repo

File: package.json [hunk 11]
```diff
@@ -6,8 +6,9 @@
     "types": "./WebApi.d.ts",
     "scripts": {
         "build": "node make.js build",
+        "samples": "node make.js samples && node samples/run.js",
```
[the remark this answers is not in the document]
Comment (damccorm) [reply package.json]: Was this meant to be on this PR https://github.com/Microsoft/typed-rest-client/pull/86? The react sample is in typed-rest-client and I think this one already exhibits that behavior, I might be misunderstanding the comment though.

**Pull request merged**